Type annotations in Python (#99)
stackbuilders/tutorials: tutorials/python/python-type-driven-development/tutorial.md

---
title: Type Driven Development in Python
published: 2019-08-01
tags: Python, typing, types, tdd
language: python
author-name: Carlos Villavicencio
twitter-profile: po5i
github-profile: po5i
description: Learn how to improve the software quality and readability by using type driven development in Python.
---

> In a [previous article][types-vs-tests], we discussed the benefits of working with Type Driven Development and the importance of static type checking in compilation time for strongly typed languages such as Java or Haskell. It doesn’t matter if their paradigm is object-oriented or functional,static types add a layer of quality, allows us to reduce the amount of tests we need to write, and reduces our dependency on tests.

Reviewer: @po5i I think is good that we reference a previous article but it might be an indirection for readers if we start pointing to a separate article first. Perhaps we could reorder things a little bit since on the second sentence you emphasize the importance of the safety provided by the static analysis which I think is a good start point to engage readers since the beginning. What do you think?

Additionally a couple of minor suggestions related to this paragraph:

```diff
- doesn't matter if their paradigm
+ doesn't matter if the paradigm
```

```diff
- static types add a layer of quality, allows us to reduce the amount
+ type-checking adds a layer of safety, reducing the amount
```

Reviewer: @po5i `and reduces our dependency on tests` I think this is a good hypothesis, maybe we should take that into account during the conclusions, for instance from the reader standpoint I would be interested in learning more about what kind of specific test I could get rid off by adding types to my code.

Type Driven Development can increase the performance, accuracy, and readability of your code by taking advantage of the language’s type system. There are many books that you can use to learn how to achieve this, for example, “Type-Driven Development with Idris” ([Manning][tdd-idris]) and extrapolate these practices to other languages.

## What about Python?

Is it possible to use a strong type system in a language like Python? The answer is yes. [PEP-484][pep-484] introduced _Type Annotations to Python_ 3.5 back in 2014. It’s been a while, right? Why this isn’t popular? Because Python doesn’t require the code has types everywhere to run, they’re completely optional and mostly used by editors and IDEs. Another reason is that Python code is well-known to be readable in the least number of possible lines of code, but adding types to your code doesn’t make it more readable? What a paradox!

> In this article, we’ll be walking through the Python type system, how to add them to our code and Docstrings and perform static type checking using [mypy][mypy]. To start using mypy, install it using pip globally or in your virtual environment.

Reviewer: @po5i `perform static type checking` I'm curious about `mypy` is it a static analysis tool, a Python superset, or just a library that provide type-checking at runtime? According to the website:

> Compile-time type checking
> Static typing makes it easier to find bugs with less debugging.

So I think it is similar to TS in that regard? I think it would be interesting to compare it with other similar tools out in the development ecosystem. What do you think?

Author: @sestrella so `mypy` falls in the first category, a static analysis library. It's not a superset because the type annotations are provided by Python natively. Neither the type-checking at runtime, the type-checking is being executed independently (that's why it's useful for a CI task for example).

I think it's a good idea to elaborate a bit more about the purpose/description of the `mypy` tool. Will work on it.

Author: Tutorial updated. @sestrella Would you mind to check it again when you have some time?

---
title: Type Driven Development in Python
published: 2019-08-01
tags: Python, typing, types, tdd
language: python
author-name: Carlos Villavicencio
twitter-profile: po5i
github-profile: po5i

> description: Learn how to improve the software quality and readability by using type driven development in Python.

Reviewer 2: I think that this description could be better if you add more details. Also, after reading the tutorial, are the examples actually about doing type-driven development or more like how to use types in Python? If the idea is to make it more about type-driven development, perhaps a bigger example could be used to illustrate how to do that.

Author: You're right. The goal of this tutorial is to provide examples and encourage the reader on how to use types in Python instead of going deep in type-driven development. I changed a bit the introductory paragraphs to do so to avoid confusion. Thanks for your review. \cc @jsl

---

> _Type Driven Development_ can increase the performance, accuracy, and readability of your code by taking advantage of the language’s type system. There are many books that you can use to learn how to achieve this, for example, “Type-Driven Development with Idris” ([Manning][tdd-idris]) and extrapolate these practices to other languages.

Reviewer 2: I think I'd go with type-driven (instead of type driven) (here and in all places where the concept is used).

```suggestion
_Type-Driven Development_ can increase the performance, accuracy, and readability of your code by taking advantage of the language’s type system. There are many books that you can use to learn how to achieve this, for example, [Type-Driven Development with Idris][tdd-idris], and extrapolate these practices to other languages.
```

In a [previous article][types-vs-tests], we discussed the benefits of working with _Type Driven Development_ and the importance of static type checking in compilation time for strongly typed languages such as Java or Haskell. It doesn’t matter if the paradigm is object-oriented or functional, type-checking adds a layer of safety, reducing the amount of tests we need to write, and reduces our dependency on tests.

## What about Python?

> Is it possible to use a strong type system in a language like Python? The answer is yes. [PEP-484][pep-484] introduced _Type Annotations to Python_ 3.5 back in 2014. It’s been a while, right? Why this isn’t popular? Because Python doesn’t require the code has types everywhere to run, they’re completely optional and mostly used by editors and IDEs. Another reason is that Python code is well-known to be readable in the least number of possible lines of code, but adding types to your code doesn’t make it more readable? What a paradox!

Reviewer 2: ```suggestion
Is it possible to use a strong type system in a language like Python? The answer is yes. [PEP-484][pep-484] introduced _Type Annotations to Python_ 3.5 back in 2014. It’s been a while, right? Why is this not popular? Because Python doesn’t require the code has types everywhere to run, they’re completely optional and mostly used by editors and IDEs. Another reason is that Python code is well-known to be readable in the least number of possible lines of code, but adding types to your code doesn’t make it more readable. What a paradox!
```

- s/Why this isn't popular?/Why is this not popular?
- s/... doesn't make it more readable?/... doesn't make it more readable.

(Since all paragraphs are one line, it's very confusing to add a suggestion.)

> In this article, we’ll be walking through the Python type system, how to add them to our code and Docstrings and perform static type checking using [mypy][mypy]. There are plenty options for static type-checking in Python, but we'll follow [Guido van Rossum suggestion][guido-mypy].

Reviewer 2: Is it to add types? s/how to add them/how to add types

### Mypy

> Mypy is a third-party Python library that provides an optional static type check. Opposite of other non-dynamic programming languages like Java, where the static type-check takes place at compilation time, Mypy CLI does the type-check to a file (or a set of files), so it's up to you to decide where in your workflow the type-checking will be executed, it can be on CI or a pre-commit hook, just to mention some.

Reviewer 2: ```suggestion
Mypy is a third-party Python library that provides optional static type checking. Unlike other non-dynamic programming languages like Java, where static type-checking takes place at compilation time...
```

I don't understand the idea here or why Mypy CLI is the reason why it's different. Can you rephrase that? Is the case that Mypy provides a command to type check a file that you'll have to manually call to trigger type checking?

Reviewer 2: Also, is Pip already linked somewhere or should a version of that be mentioned?

Reviewer 2: One more taking into account that this is a tutorial. Maybe it would help if there's an example (it could be a separate repository) with a small Python project that has this configured in CI or in whatever way you'd recommend.

Author: In the previous paragraph it's explaned why we're choosing `mypy` instead of any other option available. I also rephrased the Pip-related line.

> One more taking into account that this is a tutorial. Maybe it would help if there's an example (it could be a separate repository) with a small Python project that has this configured in CI or in whatever way you'd recommend.

Good idea, I will think about expanding the examples to a bigger one so I can cover TypeDD better, like you suggested in the last comment of your review.

To start using mypy, install it using pip globally or in your virtual environment.

```bash
pip install mypy
```

To perform static type checking to a source file:

```bash
mypy my_file.py
```

## Primitive types

> Python is a multi-paradigm dynamic language, so the type of a variable is determined based on the value.  Let's no get confused. Saying that Python is a dynamic language doesn't mean that has _weak typing_. Python is _strong typed_. Surprised? Check this out:

Reviewer 2: Something has to be fixed in "let's no get confused". Maybe "let's not get confused by this" or a different way of saying it. Or something like this:

```suggestion
Python is a multi-paradigm dynamic language, so the type of a variable is determined based on the value. This might be confusing, but it doesn't mean that Python has _weak typing_. Python is _strong typed_. Surprised? Check this out:
```

```python
movie = "Die Hard"
movie = movie + 2
# TypeError: cannot concatenate 'str' and 'int' objects
```

> That said, Let’s begin this tutorial with this basic example:

Reviewer 2: ```suggestion
That said, let’s begin this tutorial with a basic example:
```

```

If we execute this file we will expect the same output as the previous one, but when we run `mypy` we will get the following message:

```
error: Incompatible types in assignment (expression has type "float", variable has type "str")
```

Did you catch the reason? The variable `weight_pounds` was defined as `str` but we were assigning a float number to it. We can start spotting bugs in every type check.

I’m pretty sure you can understand this function:

```python
def make_hamburger(meat, number_of_meats):
   return ["bread"] + [meat] * number_of_meats + ["bread"]

print(make_hamburger("ground beef", 2))
# ['bread', 'ground beef', 'ground beef', 'bread']
```

Let’s see how it goes with types:

```python
from typing import List

def make_hamburger(meat: str, number_of_meats: int) -> List[str]:
   return ["bread"] + [meat] * number_of_meats + ["bread"]
```

What happened here? We defined the function’s arguments types: string for `meat`, integer for `number_of_meats` and the return type which is a list of string values. Did you see the import at the beginning? Complex types must be imported from the `typing` library. We will cover some of them later.

## Type Alias

> You can also define your custom type names for known structures, very useful to improve the readability. Let’s say if we want to consider a hamburger is a list of strings, we can define a type `Hamburger` in the following way:

Reviewer 2: ```suggestion
You can also define your custom type names for known structures, which is very useful for improving readability. Let’s say if we want to consider a hamburger is a list of strings, we can define a type `Hamburger` in the following way:
```

```python
Hamburger = List[str]

def make_hamburger(meat: str, number_of_meats: int) -> Hamburger:
   return ["bread"] + [meat] * number_of_meats + ["bread"]
```

## Callables

> Are you doing functional programming? No problem, the typing library inclues a type for callables which accepts a two-dimensional list like `[[argument1_type, … argumentN_type], return_type]`.

Reviewer 2: Is callable something that should be clear at this point from Python or would that require additional explanation?

Author: Good call. I'll add a brief description at the beginning.

```python
from typing import Callable, Optional

def sum_and_process(a: int, b: int, callback: Callable[[int, Optional[str]], bool]) -> bool:
   total = a + b
   return callback(total, None)

def is_positive(val: int, message: Optional[str]) -> bool:
   if message:
       print(message)
   return val > 0

output = sum_and_process(5, 2, is_positive)
print(output)
# True
```

> First of all, we are creating a function that takes two integer values, sums up them and returns the output of an incoming callback. This callback `is_positive` is defined as a function that takes one mandatory integer and one optional string. If the message exists it will print it, and always returns if the incoming value is positive. Of course, this function is not pure because of the print side-effect, but it’s just an example :)

Reviewer 2: s/sums up them/sums them up

## Generics and Union Types

One of the most powerful features in a type system is the generics and union types. It’s also possible to use these in Python annotations. Let’s take a look at the following example:

```python
from typing import TypeVar, List

T = TypeVar("T", int, List[str])

def generic_add(x: T, y: T) -> T:
   return x + y

x1: int = 5
y1: int = 2
print(generic_add(x1, y1))
# 7

x2: List[str] = ["Hello"]
y2: List[str] = ["World"]
print(generic_add(x2, y2))
# ['Hello', 'World']

x3: str = "foo"
y3: str = "bar"
print(generic_add(x3, y3))
# error: Value of type variable "T" of "generic_add" cannot be "str"
```

> Above, we defined a type variable `T` that can be integer or List of strings and a `generic_add` function that perform the “addition” operation for the incoming arguments which are limited by the `TypeVar` declaration.

Reviewer 2: I think that this paragraph could say something more about the example or at least about he error at the end. Right now it looks like it only mentions the beginning of the example.

Also, we can make use of union types. Let’s suppose we want to support any number that can only be of types integer and float:

```python
from typing import Union

Number = Union[float, int]

def union_add(x: Number, y: Number) -> Number:
   return x + y

x1: int = 5
y1: int = 2
print(union_add(x1, y1))
# 7

x2: float = 3.14
y2: float = 3.14
print(union_add(x2, y2))
# 6.28

x3: str = "2"
y3: str = "1"
print(union_add(x3, y3))
# error: Argument 1 to "union_add" has incompatible type "str"; expected "Union[float, int]"
# error: Argument 2 to "union_add" has incompatible type "str"; expected "Union[float, int]"
```

> To wrap up, static type checking will always be important to prevent bugs or implementation misuse, and allows us to take one of the main advantages of _Type Driven Development_, don't rely too much in unit tests that check the values consistency like other dynamic and weak languages do. Now you can start doing type checking automatically in your local environments or as a continuous integration job along with the rest of your pipeline.

Reviewer 2: Perhaps something nice here would be saying more about the "don't rely too much" part, that is, it's not about one or the other, but both and in this case adding this to Python can improve things so that unit tests can be very useful for something specific, and the new things here can be useful for something else that complements the other one.

Author: Section expanded.

Also, don't forget to check out the official Python [documentation for typings][python-docs]. There you will find more useful stuff such as `Tuples`, `IO`, `Generator`, `Iterable` and `Any` (_spoiler alert:_ don’t use `Any`).

> Thanks for reading this article. We hope you found this article useful.

Reviewer 2: Do tutorials have a generic message that covers this part or was that removed?

Author: True. I'll remove this line.

---
title: Type-Driven Development in Python
published: 2019-08-01
tags: Python, typing, types, tdd
language: python
author-name: Carlos Villavicencio
twitter-profile: po5i
github-profile: po5i

> description: Learn how to improve the software quality and readability in our Python code by using the built-in typing system and complementing with other software development processes like type-driven development and test-driven development.

Reviewer 3: instead of "and complementing with other" it may read better to say, "by using the built-in typing system as it complements other"...

---

> _Type-Driven Development_ is a simple approach that states to add static types to our code that are aligned with the logic. It brings many benefits such as increasing the performance, accuracy, testability, readability, and extensibility of your code by taking advantage of the language’s type system. There are many books that you can use to learn how to achieve this, for example, [“Type-Driven Development with Idris”][tdd-idris] and extrapolate these practices to other languages.

Reviewer 3: I think we need to revise the first sentence. Perhaps something like... Type-Driven Development is a technique in which we add static types to our code in order to... 

I have to read more of this post to know what a proper lead-in would be.

> In a [previous article][types-vs-tests], we discussed the benefits of working with _Type-Driven Development_ and the importance of static type checking in compilation time for strongly typed languages such as Java or Haskell. It doesn’t matter if the paradigm is object-oriented or functional, type-checking adds a layer of safety, reducing the amount of tests we need to write, and reduces our dependency on tests.

Reviewer 3: at compilation time

> In a [previous article][types-vs-tests], we discussed the benefits of working with _Type-Driven Development_ and the importance of static type checking in compilation time for strongly typed languages such as Java or Haskell. It doesn’t matter if the paradigm is object-oriented or functional, type-checking adds a layer of safety, reducing the amount of tests we need to write, and reduces our dependency on tests.

Reviewer 3: "which reduces our dependency on tests"

> 

Reviewer 3: change first "," after functional to a hyphen since the sentence switches a bit more than the comment would indicate at that point.

## What about Python?

> Is it possible to use a strong type system in a language like Python? The answer is yes. [PEP-484][pep-484] introduced _Type Annotations to Python_ 3.5 back in 2014. It’s been a while, right? Why is this not popular? Because Python doesn’t require the code has types everywhere to run, they’re completely optional and mostly used by editors and IDEs. Another reason is that Python code is well-known to be readable in the least number of possible lines of code, but adding types to your code doesn’t make it more readable, what a paradox!

Reviewer 3: I don't quite understand the part, "Because Python doesn't require the code has types everywhere to run" - can you try re-wording and I'll take another pass at it?

> Is it possible to use a strong type system in a language like Python? The answer is yes. [PEP-484][pep-484] introduced _Type Annotations to Python_ 3.5 back in 2014. It’s been a while, right? Why is this not popular? Because Python doesn’t require the code has types everywhere to run, they’re completely optional and mostly used by editors and IDEs. Another reason is that Python code is well-known to be readable in the least number of possible lines of code, but adding types to your code doesn’t make it more readable, what a paradox!

Reviewer 3: Last sentence need some re-wording. Try making "What a paradox!" a separate sentence to start since it is pretty disconnected from first part of that sentence.

> In this article, we’ll be walking through the Python type system, how to add types to our code and Docstrings and perform static type checking using [mypy][mypy]. There are plenty options for static type-checking in Python, but in this tutorial we will follow [Guido van Rossum suggestion][guido-mypy].

Reviewer 3: change "," after type system to hyphen

### Mypy

> Mypy is a third-party Python library that provides optional static type checking. Unlike other non-dynamic programming languages like Java, where the static type-checking takes place at compilation time, Mypy CLI does the type-check to a file (or a set of files), so it's up to you to decide where in your workflow the type-checking will be executed, it can be on CI or a pre-commit hook, just to mention some.

Reviewer 3: "it can be ..." should be separate sentence.

To start using mypy, install it using any version of `pip` globally or in your virtual environment.

```bash
pip install mypy
```

> To perform static type checking to a source file:

Reviewer 3: on a source file

```bash
mypy my_file.py
```

## Primitive types

> Python is a multi-paradigm dynamic language, so the type of a variable is determined based on the value. This might be confusing, but it doesn't mean that Python has _weak typing_. Python is _strong typed_. Surprised? Check this out:

Reviewer 3: strongly-typed?

Author: Yes, this is confusing for everyone at the beginning. 

Python is a dynamic language because you can do this

```
foo = 4
```

and then foo will be an `int` in runtime. But it's also a strong typing language since you can't do this

```py
>>> foo + '2'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for +: 'int' and 'str'
```
The same way you can in JS for example (which is a weak typing language)

> Python is a multi-paradigm dynamic language, so the type of a variable is determined based on the value. This might be confusing, but it doesn't mean that Python has _weak typing_. Python is _strong typed_. Surprised? Check this out:

Reviewer 3: This is an interesting section. What do you think about this post? https://existentialtype.wordpress.com/2011/03/19/dynamic-languages-are-static-languages/

Author: Sure there is an ambiguous interpretation for dynamic and static for both language and typing. I think I should expand this section in order to explain my vision for both (similar like my comment above).

```python
movie = "Die Hard"
movie = movie + 2
# TypeError: cannot concatenate 'str' and 'int' objects
```

That said, let’s begin this tutorial with a basic example:

```python
meat = "Ground beef"
print(type(meat))
# <class 'str'>

weight_pounds = 0.5
print(type(weight_pounds))
# <class 'float'>
```

> In the above example, we are using the `type` function to inspect the type representation of any variable. What if we include types to our first example code? Let’s check the following example:

Reviewer 3: include types in our first...

```python
meat: str = "Ground beef"
weight_pounds: str = 0.5
```

> If we execute this file we will expect the same output as the previous one, but when we run `mypy` we will get the following message:

Reviewer 3: If we execute this file we expect the same output as the previous one, but we actually get the following:

```python
Hamburger = List[str]

def make_hamburger(meat: str, number_of_meats: int) -> Hamburger:
   return ["bread"] + [meat] * number_of_meats + ["bread"]
```

## Callables

A callable is anything you can call, using parenthesis, and possibly passing arguments. Callables are functions, classes, methods and even instances of classes can be callables if their class implements a `__call__()` method.

Are you doing functional programming? No problem, the typing library inclues a type for callables which accepts a two-dimensional list like `[[argument1_type, … argumentN_type], return_type]`.

```python
from typing import Callable, Optional

def sum_and_process(a: int, b: int, callback: Callable[[int, Optional[str]], bool]) -> bool:
   total = a + b
   return callback(total, None)

def is_positive(val: int, message: Optional[str]) -> bool:
   if message:
       print(message)
   return val > 0

output = sum_and_process(5, 2, is_positive)
print(output)
# True
```

> First of all, we are creating a function that takes two integer values, sums them up and returns the output of an incoming callback. This callback `is_positive` is defined as a function that takes one mandatory integer and one optional string. If the message exists it will print it, and always returns if the incoming value is positive. Of course, this function is not pure because of the print side-effect, but it’s just an example :)

Reviewer 3: Period at end of paragraph. I think that even in a blog post we don't want to rely on emojis. Do we need to make the meaning more clear?

## Generics and Union Types

> 

Reviewer 3: in-Python annotations

> One of the most powerful features in a type system is the generics and union types. It’s also possible to use these in Python annotations. Let’s take a look at the following example:

Reviewer 3: are the generic and union types

```python
from typing import TypeVar, List

T = TypeVar("T", int, List[str])

def generic_add(x: T, y: T) -> T:
   return x + y

x1: int = 5
y1: int = 2
print(generic_add(x1, y1))
# 7

x2: List[str] = ["Hello"]
y2: List[str] = ["World"]
print(generic_add(x2, y2))
# ['Hello', 'World']

x3: str = "foo"
y3: str = "bar"
print(generic_add(x3, y3))
# error: Value of type variable "T" of "generic_add" cannot be "str"
```

> Above, we defined a type variable `T` that can be integer or List of strings and a `generic_add` function that perform the “addition” operation for the incoming arguments which are limited by the `TypeVar` declaration.

Reviewer 3: function that performs (third-person conjugation)

The first two invocations will work because we're passing arguments that belong to the `TypeVar` set of types and the function will behave accordingly since both types implements the `+` operation. It will actually run for the third invocation because Python implements the add operation for strings, but mypy will raise the error as `str` is not supported by the `T` definition.

Also, we can make use of union types. Let’s suppose we want to support any number that can only be of types integer and float:

```python
from typing import Union

Number = Union[float, int]

def union_add(x: Number, y: Number) -> Number:
   return x + y

x1: int = 5
y1: int = 2
print(union_add(x1, y1))
# 7

x2: float = 3.14
y2: float = 3.14
print(union_add(x2, y2))
# 6.28

x3: str = "2"
y3: str = "1"
print(union_add(x3, y3))
# error: Argument 1 to "union_add" has incompatible type "str"; expected "Union[float, int]"
# error: Argument 2 to "union_add" has incompatible type "str"; expected "Union[float, int]"
```

> To wrap up, static type checking will always be important to prevent bugs or implementation misuse, and allows us to take one of the main advantages of _Type-Driven Development_, don't rely too much in unit tests that check the values consistency like other dynamic and weak-typing languages do. It's not about stop testing, but leave them the work of testing something specific and, in the same way, complement our projects with this set of tools. Now you can start doing type checking automatically in your local environments or as a Continuous Integration job along with the rest of your pipeline.

Reviewer 3: and it allows us to

> To wrap up, static type checking will always be important to prevent bugs or implementation misuse, and allows us to take one of the main advantages of _Type-Driven Development_, don't rely too much in unit tests that check the values consistency like other dynamic and weak-typing languages do. It's not about stop testing, but leave them the work of testing something specific and, in the same way, complement our projects with this set of tools. Now you can start doing type checking automatically in your local environments or as a Continuous Integration job along with the rest of your pipeline.

Reviewer 3: After misuse maybe a new sentence? It allows us to take advantage of Type-Driven Development without relying on an excessive amount of unit tests that check the values. How are we defining weak-typing languages? Is this what we're calling "dynamically-typed" languages?

> To wrap up, static type checking will always be important to prevent bugs or implementation misuse, and allows us to take one of the main advantages of _Type-Driven Development_, don't rely too much in unit tests that check the values consistency like other dynamic and weak-typing languages do. It's not about stop testing, but leave them the work of testing something specific and, in the same way, complement our projects with this set of tools. Now you can start doing type checking automatically in your local environments or as a Continuous Integration job along with the rest of your pipeline.

Reviewer 3: Need to re-write the sentence starting with "it's not"...

---
title: Type annotations in Python
published: 2019-08-01
tags: Python, typing, types, tdd
language: python
author-name: Carlos Villavicencio
twitter-profile: po5i
github-profile: po5i
description: Learn how to improve the software quality and readability in our Python code by using the built-in typing system as it complements other software development processes like type-driven development and test-driven development.
---

> _Type-Driven Development_ is a technique in which we plan our program development by adding types first and then write the code that satisfies that type. It brings many benefits such as increasing the robustness, accuracy, testability, readability, and extensibility of your code by taking advantage of the language’s type system. We strongly suggest as follow up reading [_Type-Driven Development with Idris_][tdd-idris] and extrapolate these practices to other languages.

Reviewer 2: Can you add more information about the link to Edwin Brady's book or move it to the end of the tutorial? I didn't immediately noticed that this link is the one being referenced below when you cite Brady.

In this article, we’ll be walking through the Python - how to add types to our code and Docstrings and perform static type checking using [mypy][mypy]. There are plenty options for static type-checking in Python, but in this tutorial we will follow [Guido van Rossum suggestion][guido-mypy].

### Mypy

Mypy is a third-party Python library that provides optional static type checking. Unlike other non-dynamic programming languages like Java, where the static type-checking takes place at compilation time, Mypy CLI does the type-check to a file (or a set of files), so it's up to you to decide where in your workflow the type-checking will be executed. It can be on CI or a pre-commit hook, just to mention some.

To start using mypy, install it using any version of `pip` globally or in your virtual environment.

```bash
pip install mypy
```

To perform static type checking on a source file:

```bash
mypy my_file.py
```

## Primitive types

> Python is a multi-paradigm dynamic language, so the type of a variable is determined based on the value. This might be confusing, but it doesn't mean that Python has _weak typing_. Python is _strongly typed_. Surprised? Check this out:

Reviewer 2: Was the conversation related to this resolved? Maybe adding a short comment about the difference between weak and strong typing could help to know what the example below is saying. Or a link to something with additional information.

# <class 'str'>

weight_pounds = 0.5
print(type(weight_pounds))
# <class 'float'>
```

In the above example, we are using the `type` function to inspect the type representation of any variable. What if we include types in our first example code? Let’s check the following example:

```python
meat: str = "Ground beef"
weight_pounds: str = 0.5
```

If we execute this file we expect the same output as the previous one, but we actually get the following:

```
error: Incompatible types in assignment (expression has type "float", variable has type "str")
```

Did you catch the reason? The variable `weight_pounds` was defined as `str` but we were assigning a float number to it. We can start spotting bugs in every type check.

I’m pretty sure you can understand this function:

```python
def make_hamburger(meat, number_of_meats):
   return ["bread"] + [meat] * number_of_meats + ["bread"]

print(make_hamburger("ground beef", 2))
# ['bread', 'ground beef', 'ground beef', 'bread']
```

Let’s see how it goes with types:

```python
from typing import List

def make_hamburger(meat: str, number_of_meats: int) -> List[str]:
   return ["bread"] + [meat] * number_of_meats + ["bread"]
```

> What happened here? We defined the function’s arguments types: string for `meat`, integer for `number_of_meats` and the return type which is a list of string values. Did you see the import at the beginning? Complex types must be imported from the `typing` library. We will cover some of them later.

Reviewer 2: Was this covered later? I think there's an example using this import, but maybe that can be explicit. I think the typing library is also mentioned at the very end of the tutorial, so that might be what confused me.

Author: I think it's better to remove this "We will cover some of them later." and elaborate more abot the complex types and the `typing` lib.

The first two invocations will work because we're passing arguments that belong to the `TypeVar` set of types and the function will behave accordingly since both types implements the `+` operation. It will actually run for the third invocation because Python implements the add operation for strings, but mypy will raise the error as `str` is not supported by the `T` definition.

Also, we can make use of union types. Let’s suppose we want to support any number that can only be of types integer and float:

```python
from typing import Union

Number = Union[float, int]

def union_add(x: Number, y: Number) -> Number:
   return x + y

x1: int = 5
y1: int = 2
print(union_add(x1, y1))
# 7

x2: float = 3.14
y2: float = 3.14
print(union_add(x2, y2))
# 6.28

x3: str = "2"
y3: str = "1"
print(union_add(x3, y3))
# error: Argument 1 to "union_add" has incompatible type "str"; expected "Union[float, int]"
# error: Argument 2 to "union_add" has incompatible type "str"; expected "Union[float, int]"
```

> To wrap up, static type checking will always be important to prevent bugs or implementation misuse. It allows us to take advantage of Type-Driven Development without relying on an excessive amount of unit tests that check the values. The difference with Test-Driven Development is that, "unlike tests, which can usually only be used to show the _presence_ of errors, types (used appropriately) can show the _absense_ of errors. But although types _reduce_ the need for tests, they rarely eliminate it entirely" (Brady, 2017, p.3)

Reviewer 2: Does this quote apply to the way that types are used in Python or is this a different context? I haven't read the type-driven development book, so just confirming taking into account that the book is about Idris. Or what does "used appropriately" mean in that quote?

Reviewer 2: Also, s/absense/absence.

Author: This quote is mostly to highlight the importance of types or type-driven development, which is now possible using Python type annotations. Indeed the books is about Idris and only the first chapter is non-Idris related but still a good reference to software design using type-driven development.

> Now you can start doing type checking automatically in your local environments or as a Continuous Integration job along with the rest of your pipeline.

Reviewer 2: I think that this paragraph is missing some context or a link or something to make it fit better with the rest of the content. It's very short right now and very different to the previous one.

Author: Moved to the MyPy section.

---
title: Type annotations in Python
published: 2019-08-01
tags: Python, typing, types, tdd
language: python
author-name: Carlos Villavicencio
twitter-profile: po5i
github-profile: po5i

> description: Learn how to improve the software quality and readability in our Python code by using the built-in typing system as it complements other software development processes like type-driven development and test-driven development.

Reviewer 3: We can remove the word, "our"

---

_Type-Driven Development_ is a technique in which we plan our program development by adding types first and then write the code that satisfies that type. It brings many benefits such as increasing the robustness, accuracy, testability, readability, and extensibility of your code by taking advantage of the language’s type system.

> In a [previous article][types-vs-tests], we discussed the benefits of working with _Type-Driven Development_ and the importance of static type checking at compilation time for strongly typed languages such as Java or Haskell. Type-checking strengthens our code, reduces the number of tests we need to write which decreases our dependency on them no matter if the paradigm is object-oriented or functional.

Reviewer 3: Type-checking strengthens our code [and] reduces the number of tests we need to write which decreases our dependency on them - no matter if the paradigm is object-oriented or functional.

> In a [previous article][types-vs-tests], we discussed the benefits of working with _Type-Driven Development_ and the importance of static type checking at compilation time for strongly typed languages such as Java or Haskell. Type-checking strengthens our code, reduces the number of tests we need to write which decreases our dependency on them no matter if the paradigm is object-oriented or functional.

Reviewer 4: One could argue that this is not correct as it is possible and easy to just trick Java's type system

> strongly typed languages such as Java

Author: Good call. Do you think removing Java will be enough? or maybe replace it with another lang?

Reviewer 4: What about a footnote acknowledging that Java might not be strongly typed as there are ways to skip the type system. But, for the sake of the tutorial, it is fine to consider it like that

Author: I think that can be confusing to the reader, mostly because we're not addressing Java or covering a type system topic. But an interesting topic that can be expanded or elaborated after.

## What about Python?

> Is it possible to use a strong type system in a language like Python? The answer is yes. [PEP-484][pep-484] introduced _Type Annotations to Python_ 3.5 back in 2014. It’s been a while, right? Why is this not popular? It's because Python being a dynamic language doesn’t require to explicitly write variable or return types, they’re completely optional and mostly used by editors and IDEs.

Reviewer 3: why is this not [more] popular

> Is it possible to use a strong type system in a language like Python? The answer is yes. [PEP-484][pep-484] introduced _Type Annotations to Python_ 3.5 back in 2014. It’s been a while, right? Why is this not popular? It's because Python being a dynamic language doesn’t require to explicitly write variable or return types, they’re completely optional and mostly used by editors and IDEs.

Reviewer 3: It's because Python[,] being a dynamic language, doesn't require [us] to explicitly write variable or return types - they're completely optional...

> In this article, we’ll be walking through the Python - how to add types to our code and Docstrings and perform static type checking using [mypy][mypy]. There are plenty options for static type-checking in Python, but in this tutorial we will follow [Guido van Rossum suggestion][guido-mypy].

Reviewer 3: Not sure what you mean here - "walking through the Python"?

Author: type system* But that text was missing. Updating it...

### Mypy

> Mypy is a third-party Python library that provides optional static type checking. Unlike other non-dynamic programming languages like Java, where the static type-checking takes place at compilation time, Mypy CLI does the type-check to a file (or a set of files) on demand. Apart of having type-checking at development time, it's helpful to also include this check automatically in the Continous Integration pipeline.

Reviewer 3: apart *from*

To start using mypy, install it using any version of `pip` globally or in your virtual environment.

```bash
pip install mypy
```

To perform static type checking on a source file:

```bash
mypy my_file.py
```

## Primitive types

> You might know that programming languages can be statically or dynamically typed. A statically typed language does type checking at compile-time while dynamically type languages do it at run-time.

Reviewer 3: while dynamically *typed* *language* *does* it at run-time.

> Another concept that is important to know before we continue is the difference between weak and strong typed languages. I short, a strong typed language has stricter rules such as variable assignment, return values and function calling while weak typed ones can produce unpredictable results.

Reviewer 3: *weakly and strongly*

> That said, Python is a multi-paradigm dynamic language, so the type of variable is determined based on the value. This might be confusing, but it doesn't mean that Python has _weak typing_. Python is _strongly typed_. Surprised? Check this out:

Reviewer 3: *weakly*

weight_pounds: str = 0.5
```

If we execute this file we expect the same output as the previous one, but we actually get the following:

```
error: Incompatible types in assignment (expression has type "float", variable has type "str")
```

Did you catch the reason? The variable `weight_pounds` was defined as `str` but we were assigning a float number to it. We can start spotting bugs in every type check.

I’m pretty sure you can understand this function:

```python
def make_hamburger(meat, number_of_meats):
   return ["bread"] + [meat] * number_of_meats + ["bread"]

print(make_hamburger("ground beef", 2))
# ['bread', 'ground beef', 'ground beef', 'bread']
```

Let’s see how it goes with types:

```python
from typing import List

def make_hamburger(meat: str, number_of_meats: int) -> List[str]:
   return ["bread"] + [meat] * number_of_meats + ["bread"]
```

What happened here? We defined the function’s arguments types: string for `meat`, integer for `number_of_meats` and the return type which is a list of string values. Did you see the import at the beginning? Complex types such as `List`, `Dict` or `Tuple` must be imported from the `typing` library.

## Type Alias

> You can also define your custom type names for known structures, which is very useful for improving readability. Let’s say if we want to consider a hamburger is a list of strings, we can define a type `Hamburger` in the following way:

Reviewer 3: if we *want to specify that a hamburger*

```python
Hamburger = List[str]

def make_hamburger(meat: str, number_of_meats: int) -> Hamburger:
   return ["bread"] + [meat] * number_of_meats + ["bread"]
```

## Callables

> A callable is anything you can call, using parenthesis, and possibly passing arguments. Callables are functions, classes, methods and even instances of classes can be callables if their class implements a `__call__()` method.

Reviewer 3: Callables can be functions, classes, methods, or even instances of classes (if they implement a `__call__()` method).

Are you doing functional programming? No problem, the typing library includes a type for callables which accepts a two-dimensional list like `[[argument1_type, … argumentN_type], return_type]`.

```python
from typing import Callable, Optional

def sum_and_process(a: int, b: int, callback: Callable[[int, Optional[str]], bool]) -> bool:
   total = a + b
   return callback(total, None)

def is_positive(val: int, message: Optional[str]) -> bool:
   if message:
       print(message)
   return val > 0

output = sum_and_process(5, 2, is_positive)
print(output)
# True
```

> First of all, we are creating a function that takes two integer values, sums them up and returns the output of an incoming callback. This callback `is_positive` is defined as a function that takes one mandatory integer and one optional string. If the message exists it will print it, and always returns if the incoming value is positive.

Reviewer 3: *The* callback `is_positive` ...

and *it* returns the value if it is greater than 0 (or null otherwise)? @po5i maybe you can clarify this a bit?

## Generics and Union Types

> One of the most powerful features in a type system are the generics and union types. It’s also possible to use these in-Python annotations. Let’s take a look at the following example:

Reviewer 3: Do you mean, *are generics* and union types?

```python
from typing import TypeVar, List

T = TypeVar("T", int, List[str])

def generic_add(x: T, y: T) -> T:
   return x + y

x1: int = 5
y1: int = 2
print(generic_add(x1, y1))
# 7

x2: List[str] = ["Hello"]
y2: List[str] = ["World"]
print(generic_add(x2, y2))
# ['Hello', 'World']

x3: str = "foo"
y3: str = "bar"
print(generic_add(x3, y3))
# error: Value of type variable "T" of "generic_add" cannot be "str"
```

> Above, we defined a type variable `T` that can be integer or List of strings and a `generic_add` function that performs the “addition” operation for the incoming arguments which are limited by the `TypeVar` declaration.

Reviewer 3: that can be *an integer*

The first two invocations will work because we're passing arguments that belong to the `TypeVar` set of types and the function will behave accordingly since both types implements the `+` operation. It will actually run for the third invocation because Python implements the add operation for strings, but mypy will raise the error as `str` is not supported by the `T` definition.

Also, we can make use of union types. Let’s suppose we want to support any number that can only be of types integer and float:

```python
from typing import Union

Number = Union[float, int]

def union_add(x: Number, y: Number) -> Number:
   return x + y

x1: int = 5
y1: int = 2
print(union_add(x1, y1))
# 7

x2: float = 3.14
y2: float = 3.14
print(union_add(x2, y2))
# 6.28

x3: str = "2"
y3: str = "1"
print(union_add(x3, y3))
# error: Argument 1 to "union_add" has incompatible type "str"; expected "Union[float, int]"
# error: Argument 2 to "union_add" has incompatible type "str"; expected "Union[float, int]"
```

> To wrap up, static type checking will always be important to prevent bugs or implementation misuse. It allows us to take advantage of Type-Driven Development without relying on an excessive amount of unit tests that check the values. The difference with Test-Driven Development is that, "unlike tests, which can usually only be used to show the _presence_ of errors, types (used appropriately) can show the _absence_ of errors. But although types _reduce_ the need for tests, they rarely eliminate it entirely" (Brady, 2017, p.3). We strongly suggest as follow up reading [_Type-Driven Development with Idris_][tdd-idris] and extrapolate these practices to Python or other languages.

Reviewer 3: Break into two sentences. *While Idris allows for more expressive types than Python, many of the concepts in that book can be utilized in Python or other languages.*

---
title: Type annotations in Python
published: 2019-08-01
tags: Python, typing, types, tdd

> language: python

Reviewer 5: "Python" P in capital letter.

author-name: Carlos Villavicencio
twitter-profile: po5i
github-profile: po5i
description: Learn how to improve the software quality and readability in Python code by using the built-in typing system as it complements other software development processes like type-driven development and test-driven development.
---

_Type-Driven Development_ is a technique in which we plan our program development by adding types first and then write the code that satisfies that type. It brings many benefits such as increasing the robustness, accuracy, testability, readability, and extensibility of your code by taking advantage of the language’s type system.

> In a [previous article][types-vs-tests], we discussed the benefits of working with _Type-Driven Development_ and the importance of static type checking at compilation time for strongly typed languages such as Haskell. Type-checking strengthens our code and reduces the number of tests we need to write which decreases our dependency on them no matter if the paradigm is object-oriented or functional.

Reviewer 5: Add a comma before which: write, which

## What about Python?

> Is it possible to use a strong type system in a language like Python? The answer is yes. [PEP-3107][pep-3107] and [PEP-484][pep-484] introduced _Type Annotations to Python_ 3.5 back in 2014. It’s been a while, right? Why is this not more popular? It's because Python, being a dynamic language doesn’t require us to explicitly write variable or return types, they’re completely optional and mostly used by editors and IDEs.

Reviewer 5: Explicitly goes at the end: to write variable or return types explicitly,

> In this article, we’ll be walking through the Python type system - how to add types to our code and Docstrings and perform static type checking using [mypy][mypy]. There are plenty options for static type-checking in Python, but in this tutorial we will follow [Guido van Rossum suggestion][guido-mypy].

Reviewer 5: Add of after plenty: plenty of options
Add a comma after tutorial: but in this tutorial,

### Mypy

Mypy is a third-party Python library that provides optional static type checking. Unlike other non-dynamic programming languages like Java, where the static type-checking takes place at compilation time, Mypy CLI does the type-check to a file (or a set of files) on demand. Apart from type-checking at development time, it's helpful to also include this check automatically in the Continuous Integration pipeline.

To start using mypy, install it using any version of `pip` globally or in your virtual environment.

```bash
pip install mypy
```

To perform static type checking on a source file:

```bash
mypy my_file.py
```

## Primitive types

You might know that programming languages can be statically or dynamically typed. A statically typed language does type checking at compile-time while dynamically typed language does it at run-time.

> Another concept that is important to know before we continue is the difference between weakly and strongly typed languages. I short, a strongly typed language has stricter rules such as variable assignment, return values and function calling while weakly typed ones can produce unpredictable results.

Reviewer 5: "I short" to "In short"
A comma after return values: return values, and 
A comma before while: calling, while

That said, Python is a multi-paradigm dynamic language, so the type of variable is determined based on the value. This might be confusing, but it doesn't mean that Python has _weakly typing_. Python is _strongly typed_. Surprised? Check this out:

```python
movie = "Die Hard"
movie = movie + 2
# TypeError: cannot concatenate 'str' and 'int' objects
```

> That said, let’s begin this tutorial with a basic example:

Reviewer 5: Try to replace "that said" to avoid starting in the same way as the previous paragraph.

```python
meat = "Ground beef"
print(type(meat))
# <class 'str'>

weight_pounds = 0.5
print(type(weight_pounds))
# <class 'float'>
```

In the above example, we are using the `type` function to inspect the type representation of any variable. What if we include types in our first example code? Let’s check the following example:

```python
meat: str = "Ground beef"
weight_pounds: str = 0.5
```

If we execute this file we expect the same output as the previous one, but we actually get the following:

```
error: Incompatible types in assignment (expression has type "float", variable has type "str")
```

> Did you catch the reason? The variable `weight_pounds` was defined as `str` but we were assigning a float number to it. We can start spotting bugs in every type check.

Reviewer 5: A comma before but: ,` but 
I guess in the last sentence it is: in every type of check or typed check?

I’m pretty sure you can understand this function:

```python
def make_hamburger(meat, number_of_meats):
   return ["bread"] + [meat] * number_of_meats + ["bread"]

print(make_hamburger("ground beef", 2))
# ['bread', 'ground beef', 'ground beef', 'bread']
```

Let’s see how it goes with types:

```python
from typing import List

def make_hamburger(meat: str, number_of_meats: int) -> List[str]:
   return ["bread"] + [meat] * number_of_meats + ["bread"]
```

> What happened here? We defined the function’s arguments types: string for `meat`, integer for `number_of_meats` and the return type which is a list of string values. Did you see the import at the beginning? Complex types such as `List`, `Dict` or `Tuple` must be imported from the `typing` library.

Reviewer 5: A comma before which: type, which

```python
Hamburger = List[str]

def make_hamburger(meat: str, number_of_meats: int) -> Hamburger:
   return ["bread"] + [meat] * number_of_meats + ["bread"]
```

## Callables

A callable is anything you can call, using parenthesis, and possibly passing arguments. Callables can be functions, classes, methods, or even instances of classes (if their class implements a `__call__` method).

Are you doing functional programming? No problem, the typing library includes a type for callables which accepts a two-dimensional list like `[[argument1_type, … argumentN_type], return_type]`.

```python
from typing import Callable, Optional

def sum_and_process(a: int, b: int, callback: Callable[[int, Optional[str]], bool]) -> bool:
   total = a + b
   return callback(total, None)

def is_positive(val: int, message: Optional[str]) -> bool:
   if message:
       print(message)
   return val > 0

output = sum_and_process(5, 2, is_positive)
print(output)
# True
```

> First of all, we are creating a function that takes two integer values, sums them up and returns the output of an incoming callback. The callback `is_positive` is defined as a function that takes one mandatory integer and one optional string. If the message exists it will print it, and it returns `true` if the incoming value is greater than `0` (or `false` otherwise).

Reviewer 5: A comma before and: sums them up, and 
A comma after exists: If the message exists,

## Generics and Union Types

> One of the most powerful features in a type system are generics and union types. It’s also possible to use these in-Python annotations. Let’s take a look at the following example:

Reviewer 5: Change "are" to "is": One of the most powerful features in a type system " is ".

```python
from typing import TypeVar, List

T = TypeVar("T", int, List[str])

def generic_add(x: T, y: T) -> T:
   return x + y

x1: int = 5
y1: int = 2
print(generic_add(x1, y1))
# 7

x2: List[str] = ["Hello"]
y2: List[str] = ["World"]
print(generic_add(x2, y2))
# ['Hello', 'World']

x3: str = "foo"
y3: str = "bar"
print(generic_add(x3, y3))
# error: Value of type variable "T" of "generic_add" cannot be "str"
```

> Above, we defined a type variable `T` that can be an integer or List of strings and a `generic_add` function that performs the “addition” operation for the incoming arguments which are limited by the `TypeVar` declaration.

Reviewer 5: List without a capital letter.

> Above, we defined a type variable `T` that can be an integer or List of strings and a `generic_add` function that performs the “addition” operation for the incoming arguments which are limited by the `TypeVar` declaration.

Reviewer 5: Add a comma after arguments: arguments, which are

> The first two invocations will work because we're passing arguments that belong to the `TypeVar` set of types and the function will behave accordingly since both types implements the `+` operation. It will actually run for the third invocation because Python implements the add operation for strings, but mypy will raise the error as `str` is not supported by the `T` definition.

Reviewer 5: Add a comma after set of types: set of types, and 
Verb conjugation (implement without S): both types implement
Try to use another word instead of implement to avoid repeated words in the same paragraph: Python implements
The last sentence "It will actually..." is too hard to read. Consider shortening the sentence or breaking it into two something like this: It will actually run for the third invocation because Python implements the add operation for strings. Still, mypy will raise the error as `str` is not supported by the `T` definition

Also, we can make use of union types. Let’s suppose we want to support any number that can only be of types integer and float:

```python
from typing import Union

Number = Union[float, int]

def union_add(x: Number, y: Number) -> Number:
   return x + y

x1: int = 5
y1: int = 2
print(union_add(x1, y1))
# 7

x2: float = 3.14
y2: float = 3.14
print(union_add(x2, y2))
# 6.28

x3: str = "2"
y3: str = "1"
print(union_add(x3, y3))
# error: Argument 1 to "union_add" has incompatible type "str"; expected "Union[float, int]"
# error: Argument 2 to "union_add" has incompatible type "str"; expected "Union[float, int]"
```

> To wrap up, static type checking will always be important to prevent bugs or implementation misuse. It allows us to take advantage of Type-Driven Development without relying on an excessive amount of unit tests that check the values. The difference with Test-Driven Development is that, "unlike tests, which can usually only be used to show the _presence_ of errors, types (used appropriately) can show the _absence_ of errors. But although types _reduce_ the need for tests, they rarely eliminate it entirely" (Brady, 2017, p.3).

Reviewer 5: "To wrap up" does not appear to be modifying the subject "static type checking". Rewrite the sentence to avoid a dangling modifier. 
Do not put comma after "is that": Test-Driven Development is that,

For a deeper reading about the Type-driven Development topic, we suggest [_Type-Driven Development with Idris_][tdd-idris]. While Idris allows for more expressive types than Python, many of the concepts in that book can be utilized in Python or other languages.

> Are you worried that the performance of the program can be affected using Type Annotations? You don't need to. Python doesn't check the types in run-time although loading the `typing` library does add a small overhead with a negligible effect.

Reviewer 5: Add a comma before although: in run-time, although

> Also, don't forget to check out the official Python [documentation for typings][python-docs]. There you will find more useful stuff such as `Tuples`, `IO`, `Generator`, `Iterable` and `Any` (_spoiler alert:_ don’t use `Any`).

Reviewer 5: Add a comma before and: ,and`Any`
